cpp/arcticdb/memory_layout.hpp

```diff
@@ -0,0 +1,338 @@
+enum class EncodedFieldType : uint8_t {
+    UNKNOWN,
+    NDARRAY,
+    DICTIONARY
+};
+
+enum class BitmapFormat : uint8_t {
+    UNKNOWN,
+    DENSE,
+    BITMAGIC,
+    ROARING
+};
+
+// Each encoded field will have zero or one shapes blocks,
+// a potentially large number of values (data) blocks, and
+// an optional sparse bitmap. The block array serves as a
+// pointer to the first block
+struct EncodedField {
+    EncodedFieldType type_ = EncodedFieldType::UNKNOWN;
+    uint16_t shapes_count_ = 0u;
+    uint16_t values_count_ = 0u;
+    uint32_t sparse_map_bytes_ = 0u;
+    uint32_t items_count_ = 0u;
+    BitmapFormat format_ = BitmapFormat::UNKNOWN;
+    std::array<Block, 1> blocks_;
+};
+
+static_assert(sizeof(EncodedField) == 60);
+
+enum class EncodingVersion : uint16_t {
+    V1 = 0,
+    V2 = 1,
+    COUNT = 2
+};
+
+constexpr static uint16_t MAGIC_NUMBER = 0xFA57;
+
+// FixedHeader is the first thing in every encoded segment,
+// and is invariant between all encodings. The encoding version
+// allows for everything after this header to potentially change,
+// but is currently not used
+struct FixedHeader {
+    std::uint16_t magic_number;
+    std::uint16_t encoding_version;
+    std::uint32_t header_bytes;
+};
+
+constexpr static std::size_t FIXED_HEADER_SIZE = sizeof(FixedHeader);
+
+// This flag indicates that a segment is a compacted set of other segments
+enum class HeaderFlag : uint8_t {
+    COMPACTED,
+};
+
+// The segment is bootstrapped by a set of uncompressed EncodedFields,
+// that describe the sets of fields referencing the segment data. The
+// location of these initial fields is described by the following
+// structure.
+struct FieldBuffer {
+    mutable uint32_t fields_bytes_ = 0U;
+    mutable uint32_t offset_bytes_ = 0U;
+};
+
+// HeaderData describes the format of the data contained within
+// the segment. At the moment there are two encoding versions, a
+// legacy encoding utilizing a protobuf header, and the binary
+// encoding described by the MemoryLayout structure below.
+struct HeaderData { ;
+    EncodingVersion encoding_version_ = EncodingVersion::V1;
+    uint16_t fields_ = 0U;
+    uint8_t flags_ = 0U;
+    uint64_t footer_offset_ = 0U;
+    FieldBuffer field_buffer_;
+};
+
+// Indicates the sortedness of this segment
+enum class SortedValue : uint8_t {
+    UNKNOWN = 0,
+    UNSORTED = 1,
+    ASCENDING = 2,
+    DESCENDING = 3,
+};
+
+// Dynamic schema frames can change their schema over time,
+// adding and removing columns and changing types. A dynamic
+// schema type indicates that for each row group, not all of
+// the columns in the global descriptor will necessarily
+// be found
+enum class SchemaType : uint8_t {
+    STATIC,
+    DYNAMIC
+};
+
+// The type of indexing of a frame as a whole
+struct IndexDescriptor {
+    enum class Type : int32_t {
+        UNKNOWN = 0,
+        EMPTY = 69,
+        ROWCOUNT = 82,
+        STRING = 83,
+        TIMESTAMP = 84
```

Comment:
TIME_SYMBOL isn't used yet, it's for the cross-symbol compaction which is a bit of a work in progress. Since I'm not 100% sure about how that's going to work, I think I'd prefer adding things in later to taking them out.